### PR TITLE
Switch to using gRPC gateway for http

### PR DESCRIPTION
**Pull request by Smirl**
# Summary

- Switch to using grpc-gateway rather than custom handlers mapping between http and grpc
- I am using grpc-gateway/v2-beta.3
  - this is a pre-release version which is soon to be released
  - it is used because it supports v2 of go protobuf implementation _(see below)_
- We don't need to use the grpc-gateway as a proxy, we can use `RegisterInfrabinHandlerServer` to just call the grpc methods directly

# What is the v2 of go protobuf?

- The v1 is imported as `github.com/golang/protobuf` and implemented at https://github.com/golang/protobuf
- The v2 is imported as `google.golang.org/protobuf` and implemented at https://github.com/protocolbuffers/protobuf-go
- gRPC was removed from the `protoc-gen-go` binary into `protoc-gen-go-grpc`
- `protoc-gen-go-grpc` isn't released yet so we use the version in master
- `protoc-gen-go-grpc` requires a small change to the `InfrabinService` (i.e. embedding `UnimplementedInfrabinServer`)

### CODE REVIEW

File: internal/helpers/tools.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// +build tools
```
Comment (Smirl) [inline internal/helpers/tools.go:1]: This file is recommended to be able to pin versions of the tools in `go.mod`.
This is why in github-actions and `Dockerfile` we can just do `go mod tidy` to resolve / fetch the versions, then `go install ....` to ensure we all have the same versions.

File: pkg/infrabin/http.go [hunk 2]
```diff
@@ -2,72 +2,13 @@ package infrabin
```
Comment (Smirl) [inline pkg/infrabin/http.go:2]: Probably easier to see this side by side as it has completely changed.

File: pkg/infrabin/infrabin.go [hunk 3]
```diff
@@ -23,6 +23,7 @@ import (
 const AWS_METADATA_ENDPOINT = "http://169.254.169.254/latest/meta-data/"
 
 type InfrabinService struct{
+	UnimplementedInfrabinServer
```
Comment (Smirl) [inline pkg/infrabin/infrabin.go:26]: This is required for using v2 go protobuf implementation

File: proto/google/api/annotations.proto [hunk 4]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) 2015, Google Inc.
```
Comment (Smirl) [inline proto/google/api/annotations.proto:1]: I have checked in all of the protos and followed the structure as it recommends

File: proto/infrabin.proto [hunk 5]
```diff
@@ -3,14 +3,57 @@ syntax = "proto3";
...
+    }
+    rpc AWS(AWSRequest) returns (google.protobuf.Struct) {
+        option (google.api.http) = {
+            get: "/aws/{path=**}"
+        };
+    }
+
+    rpc Root(Empty) returns (Response) {
+        option (google.api.http) = {
+            get: "//"
```
Comment (Smirl) [inline proto/infrabin.proto:48]: As per https://github.com/grpc-ecosystem/grpc-gateway/issues/1500

File: README.md [hunk 6]
```diff
@@ -9,97 +9,163 @@
...
+* _grpc_: `infrabin.Infrabin.AWS` _rest_: `GET /aws/<path>`
+    * _NOTE_: `--enable-proxy-endpoint` must be set
+    * _grpc request_
+        ```
+        message AWSRequest {
+            string path = 1;
+        }
+        ```
+    * _returns_:
+        json of AWS GET call
```
Comment (maruina) [inline README.md:119]:
```suggestion
        JSON of AWS GET call
```

File: README.md [hunk 7]
```diff
@@ -9,97 +9,163 @@
...
+        ```
+        message ProxyRequest {
+            string method = 1;
+            string url = 2;
+            google.protobuf.Struct body = 3;
+            map<string, string> headers = 4;
+        }
+        ```
+    * _returns_:
+        json of proxied request
```
Comment (maruina) [inline README.md:108]:
```suggestion
        JSON of proxied request
```

**Pull request merged**